docs/developers/runtime_abstraction_layer.md

```diff
@@ -0,0 +1,208 @@
+template <typename T>
+void gemm(RalContext*, gpu_stream_handle*, MemRef<T, 2> lhs, MemRef<T, 2> rhs, MemRef<T, 2> out, bool lhs_transposed, bool, rhs_transposed);
+
+// ----====================================
+
+// register a new RAL API using macro `TAO_RAL_API`.
+//   Usage: TAO_RAL_API(api_prefix, device, api_function);
+//   - generate a unique name based on the prefix provided and the types of inputs and outputs of the function.
+//   - automatically provide a uniform interface wrapper
+//   - register the wrapper function using the unique name generated.
+
+// register ral_gemm with float type
+TAO_RAL_API("ral_gemm", "cpu", gemm<float>)
+// register ral_gemm with half type
+TAO_RAL_API("ral_gemm", "gpu", gemm<half>)
+```
+
+Thus, based on the mechanism in compiler side and the mechanism in c++ side,
+users don't need to worry about the details of c interface implementation and
+can use the c++ level api directly.
+
+## 2.2，From the Perspective of Runtime Side
+
+### 2.2.1 Overview of the Runtime Side of RAL
+
+The runtime side of the RAL can be divided into three major parts as shown in
+the following ![diagram](./RAL.png).
+
+* Interacting with the host environments (e.g. Tensorflow, PyTorch). This part
+  ususlly includes i/o bindings and other bookkeeping stuff. We need different
+  implementation of the part for different hosting environments.
+* Providing an implementation of the initialization-free driver api of each kind
+  of device. In RAL, each kind of device defines its own core set of driver
+  apis. The driver ususlly includes memory management, task launching and
+  synchronizaiton. It's worth noting that we need different driver
+  implementations for the same device type in different host environments. One
+  reason is that different host environment usually have different device
+  abstraction layer, leading to different api style and semantics. Another
+  reason is that using host environment native device resource access style
+  could also minimize the overhead when crossing host enviroment and RAL
+  boundary.
+* Providing an implementation of the custom kernel. Examples are vendor library
+  kernel, RNG, TopK or sort op. These kernels are ususlly not codegen by
+  BladeDISC and rely on some mannual crafted implementations. It worth noting
+  that the custom kernel itself uses the ral driver api to access device
+  resoruces, and thus can be shared among different host environments since the
+  ral dirver has already hided the differences among different host
+  environments.
+
+### 2.2.2 RAL Context
+
+We provide different RAL Context implementation for different target host
+environments. The context implementation consists of the above three major
+parts. Currently we have tf context implementation, which is used to be
+integrated with tensorflow (both cpu & gpu), and base context implmentation,
+which is used to be integrated with Pytorch and sometimes the standalone binary.
```

Comment:
Nit. For literal "Pytorch", I think either "PyTorch" or "Pytorch" should be fine, assuming that we use it in a consistent approach.